// eslint-plugin-react/lib/rules/destructuring-assignment.js
/**
 * @fileoverview Enforce consistent usage of destructuring assignment of props, state, and context.
 */

'use strict';

const Components = require('../util/Components');
const docsUrl = require('../util/docsUrl');
const eslintUtil = require('../util/eslint');
const isAssignmentLHS = require('../util/ast').isAssignmentLHS;
const report = require('../util/report');

const getScope = eslintUtil.getScope;
const getText = eslintUtil.getText;

const DEFAULT_OPTION = 'always';

function createSFCParams() {
  const queue = [];

  return {
    push(params) {
      queue.unshift(params);
    },
    pop() {
      queue.shift();
    },
    propsName() {
      const found = queue.find((params) => {
        const props = params[0];
        return props && !props.destructuring && props.name;
      });
      return found && found[0] && found[0].name;
    },
    contextName() {
      const found = queue.find((params) => {
        const context = params[1];
        return context && !context.destructuring && context.name;
      });
      return found && found[1] && found[1].name;
    },
  };
}

function evalParams(params) {
  return params.map((param) => ({
    destructuring: param.type === 'ObjectPattern',
    name: param.type === 'Identifier' && param.name,
  }));
}

const messages = {
  noDestructPropsInSFCArg: 'Must never use destructuring props assignment in SFC argument',
  noDestructContextInSFCArg: 'Must never use destructuring context assignment in SFC argument',
  noDestructAssignment: 'Must never use destructuring {{type}} assignment',
  useDestructAssignment: 'Must use destructuring {{type}} assignment',
  destructureInSignature: 'Must destructure props in the function signature.',
};

/** @type {import('eslint').Rule.RuleModule} */
module.exports = {
  meta: {
    docs: {
      description: 'Enforce consistent usage of destructuring assignment of props, state, and context',
      category: 'Stylistic Issues',
      recommended: false,
      url: docsUrl('destructuring-assignment'),
    },
    fixable: 'code',
    messages,

    schema: [{
      type: 'string',
      enum: [
        'always',
        'never',
      ],
    }, {
      type: 'object',
      properties: {
        ignoreClassFields: {
          type: 'boolean',
        },
        destructureInSignature: {
          type: 'string',
          enum: [
            'always',
            'ignore',
          ],
        },
      },
      additionalProperties: false,
    }],
  },

  create: Components.detect((context, components, utils) => {
    const configuration = context.options[0] || DEFAULT_OPTION;
    const ignoreClassFields = (context.options[1] && (context.options[1].ignoreClassFields === true)) || false;
    const destructureInSignature = (context.options[1] && context.options[1].destructureInSignature) || 'ignore';
    const sfcParams = createSFCParams();

    /**
     * @param {ASTNode} node We expect either an ArrowFunctionExpression,
     *   FunctionDeclaration, or FunctionExpression
     */
    function handleStatelessComponent(node) {
      const params = evalParams(node.params);

      const SFCComponent = components.get(getScope(context, node).block);
      if (!SFCComponent) {
        return;
      }
      sfcParams.push(params);

      if (params[0] && params[0].destructuring && components.get(node) && configuration === 'never') {
        report(context, messages.noDestructPropsInSFCArg, 'noDestructPropsInSFCArg', {
          node,
        });
      } else if (params[1] && params[1].destructuring && components.get(node) && configuration === 'never') {
        report(context, messages.noDestructContextInSFCArg, 'noDestructContextInSFCArg', {
          node,
        });
      }
    }

    function handleStatelessComponentExit(node) {
      const SFCComponent = components.get(getScope(context, node).block);
      if (SFCComponent) {
        sfcParams.pop();
      }
    }

    function handleSFCUsage(node) {
      const propsName = sfcParams.propsName();
      const contextName = sfcParams.contextName();
      // props.aProp || context.aProp
      const isPropUsed = (
        (propsName && node.object.name === propsName)
          || (contextName && node.object.name === contextName)
      )
        && !isAssignmentLHS(node);
      if (isPropUsed && configuration === 'always' && !node.optional) {
        report(context, messages.useDestructAssignment, 'useDestructAssignment', {
          node,
          data: {
            type: node.object.name,
          },
        });
      }
    }

    function isInClassProperty(node) {
      let curNode = node.parent;
      while (curNode) {
        if (curNode.type === 'ClassProperty' || curNode.type === 'PropertyDefinition') {
          return true;
        }
        curNode = curNode.parent;
      }
      return false;
    }

    function handleClassUsage(node) {
      // this.props.Aprop || this.context.aProp || this.state.aState
      const isPropUsed = (
        node.object.type === 'MemberExpression' && node.object.object.type === 'ThisExpression'
        && (node.object.property.name === 'props' || node.object.property.name === 'context' || node.object.property.name === 'state')
        && !isAssignmentLHS(node)
      );

      if (
        isPropUsed && configuration === 'always'
        && !(ignoreClassFields && isInClassProperty(node))
      ) {
        report(context, messages.useDestructAssignment, 'useDestructAssignment', {
          node,
          data: {
            type: node.object.property.name,
          },
        });
      }
    }

    // valid-jsdoc cannot read function types
    // eslint-disable-next-line valid-jsdoc
    /**
     * Find a parent that satisfy the given predicate
     * @param {ASTNode} node
     * @param {(node: ASTNode) => boolean} predicate
     * @returns {ASTNode | undefined}
     */
    function findParent(node, predicate) {
      let n = node;
      while (n) {
        if (predicate(n)) {
          return n;
        }
        n = n.parent;
      }
      return undefined;
    }

    return {

      FunctionDeclaration: handleStatelessComponent,

      ArrowFunctionExpression: handleStatelessComponent,

      FunctionExpression: handleStatelessComponent,

      'FunctionDeclaration:exit': handleStatelessComponentExit,

      'ArrowFunctionExpression:exit': handleStatelessComponentExit,

      'FunctionExpression:exit': handleStatelessComponentExit,

      MemberExpression(node) {
        const SFCComponent = utils.getParentStatelessComponent(node);
        if (SFCComponent) {
          handleSFCUsage(node);
        }

        const classComponent = utils.getParentComponent(node);
        if (classComponent) {
          handleClassUsage(node);
        }
      },

      TSQualifiedName(node) {
        if (configuration !== 'always') {
          return;
        }
        // handle `typeof props.a.b`
        if (node.left.type === 'Identifier'
          && node.left.name === sfcParams.propsName()
          && findParent(node, (n) => n.type === 'TSTypeQuery')
          && utils.getParentStatelessComponent(node)
        ) {
          report(context, messages.useDestructAssignment, 'useDestructAssignment', {
            node,
            data: {
              type: 'props',
            },
          });
        }
      },

      VariableDeclarator(node) {
        const classComponent = utils.getParentComponent(node);
        const SFCComponent = components.get(getScope(context, node).block);

        const destructuring = (node.init && node.id && node.id.type === 'ObjectPattern');
        // let {foo} = props;
        const destructuringSFC = destructuring && (node.init.name === 'props' || node.init.name === 'context');
        // let {foo} = this.props;
        const destructuringClass = destructuring && node.init.object && node.init.object.type === 'ThisExpression' && (
          node.init.property.name === 'props' || node.init.property.nameect.assign({method:c,headers:f,body:p},d)),m=await this.send(w,r);return d.plainText||null!=(o=m.headers.get("content-type"))&&o.includes("text")?await m.text():await m.json()}extend(e){return new R(e,this.opts)}}const q=new R({}),O=["method","query","path","service","data"];class E extends R{constructor(e={},t=i){super(e,t)}sendProxyRequest(e){const{method:t,query:s={},path:o,service:r,data:i}=e,a=(0,n.Z)(e,O);if(!o)throw new Error("Path is required");const c=d(s),u=null!=c&&c.toString()?o+"?"+c.toString():o,l=Object.assign({path:u,service:r,method:t},i);return super.request(Object.assign({},a,{data:l,method:"post"}))}request(e){const t=g(this.opts,e);this.validateOptions(t);const{url:s}=t;return/\/ws\/proxy/.test(s.toString())?this.sendProxyRequest(t):super.request(t)}extend(e){return new E(e,this.opts)}}const S=new E({});var T=s(540);const j=(e=q.request)=>(t,s=[])=>{const n=(0,T.useRef)(new AbortController);(0,T.useEffect)((()=>{const e=t.signal,s=()=>{n.current.abort()};return e instanceof AbortSignal&&e.addEventListener("abort",s),()=>{null==e||e.removeEventListener("abort",s)}}),[t.signal]);const[o,r]=(0,T.useReducer)(((e,t)=>{switch(t.type){case"loading":return Object.assign({},e,{loading:!0});case"fetched":return Object.assign({},e,{loading:!1,data:t.payload,error:void 0});case"error":return Object.assign({},e,{loading:!1,data:void 0,error:t.payload});default:return e}}),{loading:!0,data:void 0,error:void 0});return(0,T.useEffect)((()=>{const s=n.current;return(async()=>{r({type:"loading"});try{const o=await e(Object.assign({},t,{signal:n.current.signal}));if(s.signal.aborted)return;r({type:"fetched",payload:o})}catch(e){if(s.signal.aborted)return;r({type:"error",payload:e})}})(),()=>{s.abort()}}),s),o},P=j(q.request.bind(q)),A=j(S.request.bind(S));var C=s(4021),x=s(8160);function k(e,t){if(!Object.prototype.hasOwnProperty.call(e,t))throw new TypeError("attempted to use private field on non-instance");return e}var L=0;function N(e){return"__private_"+L+++"_"+e}var U=s(447
