// tsconfig-paths/lib/__tests__/tsconfig-loader.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tsconfig_loader_1 = require("../tsconfig-loader");
var path_1 = require("path");
describe("tsconfig-loader", function () {
    it("should find tsconfig in cwd", function () {
        var result = (0, tsconfig_loader_1.tsConfigLoader)({
            cwd: "/foo/bar",
            getEnv: function (_) { return undefined; },
            loadSync: function (cwd) {
                return {
                    tsConfigPath: "".concat(cwd, "/tsconfig.json"),
                    baseUrl: "./",
                    paths: {},
                };
            },
        });
        // assert.equal(result.tsConfigPath, "/foo/bar/tsconfig.json");
        expect(result.tsConfigPath).toBe("/foo/bar/tsconfig.json");
    });
    it("should return loaderResult.tsConfigPath as undefined when not found", function () {
        var result = (0, tsconfig_loader_1.tsConfigLoader)({
            cwd: "/foo/bar",
            getEnv: function (_) { return undefined; },
            loadSync: function (_) {
                return {
                    tsConfigPath: undefined,
                    baseUrl: "./",
                    paths: {},
                };
            },
        });
        // assert.isUndefined(result.tsConfigPath);
        expect(result.tsConfigPath).toBeUndefined();
    });
    it("should use TS_NODE_PROJECT env if exists", function () {
        var result = (0, tsconfig_loader_1.tsConfigLoader)({
            cwd: "/foo/bar",
            getEnv: function (key) {
                return key === "TS_NODE_PROJECT" ? "/foo/baz" : undefined;
            },
            loadSync: function (cwd, fileName) {
                if (cwd === "/foo/bar" && fileName === "/foo/baz") {
                    return {
                        tsConfigPath: "/foo/baz/tsconfig.json",
                        baseUrl: "./",
                        paths: {},
                    };
                }
                return {
                    tsConfigPath: undefined,
                    baseUrl: "./",
                    paths: {},
                };
            },
        });
        // assert.equal(result.tsConfigPath, "/foo/baz/tsconfig.json");
        expect(result.tsConfigPath).toBe("/foo/baz/tsconfig.json");
    });
    it("should use TS_NODE_BASEURL env if exists", function () {
        var result = (0, tsconfig_loader_1.tsConfigLoader)({
            cwd: "/foo/bar",
            getEnv: function (key) {
                return key === "TS_NODE_BASEURL" ? "SOME_BASEURL" : undefined;
            },
            loadSync: function (_0, _1, baseUrl) {
                return {
                    tsConfigPath: undefined,
                    baseUrl: baseUrl,
                    paths: {},
                };
            },
        });
        // assert.equal(result.baseUrl, "SOME_BASEURL");
        expect(result.baseUrl).toBe("SOME_BASEURL");
    });
    it("should not use TS_NODE_BASEURL env if it does not exist", function () {
        var result = (0, tsconfig_loader_1.tsConfigLoader)({
            cwd: "/foo/bar",
            getEnv: function (_) {
                return undefined;
            },
            loadSync: function (_0, _1, baseUrl) {
                return {
                    tsConfigPath: undefined,
                    baseUrl: baseUrl,
                    paths: {},
                };
            },
        });
        // assert.equal(result.baseUrl, undefined);
        expect(result.baseUrl).toBeUndefined();
    });
});
describe("walkForTsConfig", function () {
    it("should find tsconfig in starting directory", function () {
        var pathToTsconfig = (0, path_1.join)("/root", "dir1", "tsconfig.json");
        var res = (0, tsconfig_loader_1.walkForTsConfig)((0, path_1.join)("/root", "dir1"), function (path) { return path === pathToTsconfig; });
        // assert.equal(res, pathToTsconfig);
        expect(res).toBe(pathToTsconfig);
    });
    it("should find tsconfig in parent directory", function () {
        var pathToTsconfig = (0, path_1.join)("/root", "tsconfig.json");
        var res = (0, tsconfig_loader_1.walkForTsConfig)((0, path_1.join)("/root", "dir1"), function (path) { return path === pathToTsconfig; });
        // assert.equal(res, pathToTsconfig);
        expect(res).toBe(pathToTsconfig);
    });
    it("should return undefined when reaching the top", function () {
        var res = (0, tsconfig_loader_1.walkForTsConfig)((0, path_1.join)("/root", "dir1", "kalle"), function () { return false; });
        // assert.equal(res, undefined);
        expect(res).toBeUndefined();
    });
});
describe("loadConfig", function () {
    it("should load a config", function () {
        var config = { compilerOptions: { baseUrl: "hej" } };
        var res = (0, tsconfig_loader_1.loadTsconfig)("/root/dir1/tsconfig.json", function (path) { return path === "/root/dir1/tsconfig.json"; }, function (_) { return JSON.stringify(config); });
        // assert.deepEqual(res, config);
        expect(res).toStrictEqual(config);
    });
    it("should load a config with comments", function () {
        var config = { compilerOptions: { baseUrl: "hej" } };
        var res = (0, tsconfig_loader_1.loadTsconfig)("/root/dir1/tsconfig.json", function (path) { return path === "/root/dir1/tsconfig.json"; }, function (_) { return "{\n          // my comment\n          \"compilerOptions\": { \n            \"baseUrl\": \"hej\"\n          }\n        }"; });
        // assert.deepEqual(res, config);
        expect(res).toStrictEqual(config);
    });
    it("should load a config with trailing commas", function () {
        var config = { compilerOptions: { baseUrl: "hej" } };
        var res = (0, tsconfig_loader_1.loadTsconfig)("/root/dir1/tsconfig.json", function (path) { return path === "/root/dir1/tsconfig.json"; }, function (_) { return "{\n          \"compilerOptions\": { \n            \"baseUrl\": \"hej\",\n          },\n        }"; });
        // assert.deepEqual(res, config);
        expect(res).toStrictEqual(config);
    });
    it("should throw an error including the file path when encountering invalid JSON5", function () {
        expect(function () {
            return (0, tsconfig_loader_1.loadTsconfig)("/root/dir1/tsconfig.json", function (path) { return path === "/root/dir1/tsconfig.json"; }, function (_) { return "{\n            \"compilerOptions\": {\n          }"; });
        }).toThrowError("/root/dir1/tsconfig.json is malformed JSON5: invalid end of input at 3:12");
    });
    it("should load a config with string extends and overwrite all options", function () {
        var firstConfig = {
            extends: "../base-config.json",
            compilerOptions: { baseUrl: "kalle", paths: { foo: ["bar2"] } },
        };
        var firstConfigPath = (0, path_1.join)("/root", "dir1", "tsconfig.json");
        var baseConfig = {
            compilerOptions: {
                baseUrl: "olle",
                paths: { foo: ["bar1"] },
                strict: true,
            },
        };
        var baseConfigPath = (0, path_1.join)("/root", "base-config.json");
        var res = (0, tsconfig_loader_1.loadTsconfig)((0, path_1.join)("/root", "dir1", "tsconfig.json"), function (path) { return path === firstConfigPath || path === baseConfigPath; }, function (path) {
            if (path === firstConfigPath) {
                return JSON.stringify(firstConfig);
            }
            if (path === baseConfigPath) {
                return JSON.stringify(baseConfig);
            }
            return "";
        });
        // assert.deepEqual(res, {
        //   extends: "../base-config.json",
        //   compilerOptions: {
        //     baseUrl: "kalle",
        //     paths: { foo: ["bar2"] },
        //     strict: true,
        //   },
        // });
        expect(res).toEqual({
            extends: "../base-config.json",
            compilerOptions: {
                baseUrl: "kalle",
                paths: 4636 7.99996L9.99992 3.64641L10.707 4.35352L7.06058 7.99996Z",fill:"#6B6B6B"})),L=l().forwardRef((function(e,t){return l().createElement(w.Z,(0,m.Z)({},e,{svg:S,ref:t}))}));L.getName=function(){return"icon-direction-left"};const N=L;var P=s(3341),j=s(1730),M=s(7569);const H="horizontal",D="vertical",B="hidden",Z="down",O="left",E="right";var X,Y;const W="panes-resizer-stylesheet";class _ extends l().Component{render(){let e,t=this.props.defaultLayout||B;if(this.props.isCollapsedX&&(t=D),this.props.isCollapsedY&&(t=H),t===H?e=this.props.horizontalComponent&&l().cloneElement(this.props.horizontalComponent,{isCollapsed:this.props.isCollapsedY,onCollapseToggle:this.props.onCollapseToggle&&this.props.onCollapseToggle.bind(this,this.props.layoutAxis||"y")}):t===D&&(e=this.props.verticalComponent&&l().cloneElement(this.props.verticalComponent,{isCollapsed:this.props.isCollapsedX,onCollapseToggle:this.props.onCollapseToggle&&this.props.onCollapseToggle.bind(this,this.props.layoutAxis||"x")})),!e)return null;const s=v()({"pane-header":!0,[t]:t},this.props.className);return(0,r.jsx)("div",{className:s,style:this.props.style,"data-testid":this.props["data-testid"],children:e})}}_.propTypes={defaultLayout:y().oneOf([H,D,B]),isCollapsedX:y().bool,isCollapsedY:y().bool,onCollapseToggle:y().func,horizontalComponent:y().element,verticalComponent:y().element,className:y().string,style:y().object,layoutAxis:y().string},_.defaultProps={defaultLayout:B};let k=(0,g.observer)(X=class extends l().Component{constructor(...e){super(...e),this.handleClick=()=>(this.props.onCollapseToggle(),!1)}render(){const e=v()("pane-label",{[this.props.layout]:this.props.layout},this.props.className),t={up:R,[Z]:T.Z,[O]:N,[E]:P.Z}["function"==typeof this.props.arrowDirection?this.props.arrowDirection(this.props.isCollapsed):this.props.arrowDirection];return(0,r.jsxs)("div",{className:e,children:[(0,r.jsx)("div",{className:"pane-label__text",children:(0,r.jsx)("span",{onClick:this.props.onClick,children:this.props.label})}),this.props.actions&&(0,r.jsx)("div",{className:"pane-label__actions",children:this.props.actions}),t&&(0,r.jsx)("div",{onClick:this.handleClick,className:"pane-label__arrow","data-testid":"collapse-expand-btn",children:(0,r.jsx)(t,{color:"content-color-tertiary"})})]})}})||X;k.propTypes={label:y().oneOfType([y().element,y().string]),layout:y().oneOf([H,D,B]),arrowDirection:y().oneOfType([y().oneOf(["up",Z,O,E]),y().func]),actions:y().element,onClick:y().func,className:y().string},k.defaultProps={layout:B};let q=(0,g.observer)(Y=class extends l().Component{render(){let e=this.props.isCollapsedX||this.props.isCollapsedY,t=this.props.style||{};return e&&this.props.unmountOnCollapse?null:(e&&(t.display="none"),(0,r.jsx)("div",{className:v()("pane-content",this.props.className),style:t,children:this.props.children}))}})||Y;q.propTypes={children:y().oneOfType([y().element,y().arrayOf(y().element)]),isCollapsedX:y().bool,isCollapsedY:y().bool,className:y().string,style:y().object};class A extends l().Component{constructor(e){var t,s,i,l,n,o,a,h;super(e);const r=this.getDimensions(e);this.heights=r.height,this.widths=r.width,this.resizable={x:(null==e||null==(t=e.width)?void 0:t.resizableByDefault)||!0,y:(null==e||null==(s=e.height)?void 0:s.resizableByDefault)||!0},this.state={height:null!=(i=this.heights)&&i.collapsedByDefault?this.heights.collapsed:this.heights.default,width:null!=(l=this.widths)&&l.collapsedByDefault?this.widths.collapsed:this.widths.default,collapsedX:(null==e||null==(n=e.width)?void 0:n.collapsedByDefault)||!1,collapsedY:(null==e||null==(o=e.height)?void 0:o.collapsedByDefault)||!1,hiddenX:(null==e||null==(a=e.width)?void 0:a.hiddenByDefault)||!1,hiddenY:(null==e||null==(h=e.height)?void 0:h.hiddenByDefault)||!1},this.isResizable=this.isResizable.bind(this),this.isCollapsed=this.isCollapsed.bind(this),this.getNormalizedHeight=this.getNormalizedHeight.bind(this),this.getNormalizedWidth=this.getNormalizedWidth.bind(this),this.getMinimumUncollapsedHeight=this.getMinimumUncollapsedHeight.bind(this),this.getMinimumUncollapsedWidth=this.getMinimumUncollapsedWidth.bind(this),this.resize=this.resize.bind(this),this.collapse=this.collapse.bind(this),this.unCollapse=this.unCollapse.bind(this),this.handleCollapseToggle=this.handleCollapseToggle.bind(this),this.handleHideToggle=this.handleHideToggle.bind(this)}componentDidUpdate(e,t){var s,i;if(this.state.collapsedX&&!t.collapsedX)null==(s=(i=this.props).onCollapse)||s.call(i);else if(!this.state.collapsedX&&t.collapsedX){var l,n;null==(l=(n=this.props).onExpand)||l.call(n)}}getDimensions(e){let t=e.height||{},s=e.width||{};l().Children.forEach(e.children,(e=>{var i,l;if(e.type===_)t.collapsed=(null==(i=e.props)?void 0:i.height)||t.collapsed||0,t.default=I(t.default,t.collapsed),s.collapsed=(null==(l=e.props)?void 0:l.width)||s.collapsed||0,s.default=I(s.default,s.collapsed);else if(e.type===q){var n,o;t.default=I(t.default,(null==(n=e.props)?void 0:n.height.default)||0),s.default=I(s.default,(null==(o=e.props)?void 0:o.width.default)||0)}}));for(const e of[t,s])"number"!=typeof e.default&&(e.default=0),"number"!=typeof e.collapseThreshold&&("number"==typeof e.min?e.collapseThreshold=e.min/2:"number"==typeof e.collapsed&&(e.collapseThreshold=2*e.collapsed)),"number"!=typeof e.uncollapsed&&(e.uncollapsed=e.default),"number"==typeof e.min&&"number"==typeof e.default&&e.min>e.default&&(e.default=e.min),"number"==typeof e.max&&"number"==typeof e.default&&e.max<e.default&&(e.default=e.max),"number"==typeof e.collapsed&&"number"==typeof e.collapseThreshold&&e.collapsed>=e.collapseThreshold&&(e.collapseThreshold=e.collapsed+1);return{height:t,width:s}}isResizable(e="y"){return this.resizable[e]}isCollapsed(e="y"){return"y"===e?this.state.collapsedY:this.state.collapsedX}isHidden(e="x"){return"x"===e?this.state.hiddenX:this.state.hiddenY}setResizable(e,t=!0){this.resizable[e]=t}getNormalizedDimension(e,t,s,i,l,n,o){if("number"!=typeof n)return 0;o=o||{};const a=(e=e||{}).min,h=e.max,r=e.collapsed,d=e.collapseThreshold,p="number"==typeof r;if(l)return 0;if(!i)return p&&s?r:"number"==typeof t?t:0;if(e.shouldResistResize&&o.allowResizeResist)return"number"==typeof t?t:0;if(p&&s&&!o.shouldOpenCollapsed)return r;if(p&&o.shouldCollapseOnMin&&n<a)return r;if(p&&o.shouldCollapse){if("number"==typeof d&&n<d)return r;if(n<r)return r}else if(!o.shouldCollapse&&"number"==typeof a&&n<a)return a;return"number"==typeof a&&n<a?a:"number"==typeof h&&n>h?h:n<0?0:n}getNormalizedHeight(e,t){return this.getNormalizedDimension(this.heights,this.state.height,this.state.collapsedY,this.isResizable("y"),t
