// minimizer-webpack-plugin/dist/minify.js
ite your descriptions using the [visual editor](https://learning.postman.com/docs/publishing-your-api/authoring-your-documentation/#writing-descriptions-in-the-postman-editor) or you can enter Markdown syntax.

To add a description to a collection or folder, do the following:

1. Select the **Collections** tab, and then select a collection or folder.
1. Enter a description in the **Overview** tab.
1. Select outside of the editor to save your description.

To add a description to a request, do the following:

1. Select the **Collections** tab, and then select a request.
1. Select **View Documentation** in the top right.
1. Enter a description in the right sidebar.
1. Select outside of the editor to save your description.

To learn more about writing documentation, see [Postman's documentation](https://learning.postman.com/docs/publishing-your-api/authoring-your-documentation/).

### View documentation

You can view complete documentation for your collection that includes a description of each request and details such as the method and URL, the required authorization type, and any headers or parameters. For each request, you can view sample code in various client languages, and example response bodies and headers.

To view the complete documentation for a collection, do the following:

1. Select the **Collections** tab, then select a collection.
1. In the **Overview** tab under the **Description**, select **View complete documentation**. The VS Code extension will open the complete documentation in a new tab.

You can customize the appearance of the documentation by selecting a language to use for sample code.

## Manage environments

You can add a set of variables to an environment that you can use in your requests.

![Create new environment](https://st-ar.cdn.postman.com/release-notes/vs-code/overview/create-new-environment-in-postman-2507.gif)

To create an environment and add variables to it, do the following:

1. Select a workspace from the workspace dropdown menu in the sidebar.
1. Select the **Environments** tab.
1. To create an environment, select the **+** icon.
1. [Add variables to the environment.](https://learning.postman.com/docs/sending-requests/managing-environments/#adding-environment-variables)
1. Select **Save**.

To use environment variables in a request, do the following:

1. Select the **Collections** tab.
1. Select a request.
1. [Select an environment from the environment selector](https://learning.postman.com/docs/sending-requests/managing-environments/#selecting-an-active-environment) in the top-right corner.

To learn more about managing environments, see [Postman's documentation](https://learning.postman.com/docs/sending-requests/managing-environments/).

## Use cookies

You can view and edit cookies associated with different domains. You can then use the cookies stored in the cookie jar when sending HTTP requests in Postman.

![Manage cookies](https://st-ar.cdn.postman.com/release-notes/vs-code/overview/request-cookies-2507.gif)

To manage cookies in the VS Code extension, open an HTTP request, then select **Cookies** (below **Send**).

> You can also open the cookie editor with the `Postman: Open cookies editor` command in the [Command Palette](https://code.visualstudio.com/docs/getstarted/userinterface#_command-palette).

To learn more about using cookies, see [Postman's documentation](https://learning.postman.com/docs/sending-requests/cookies/).

## Test your API

You can use the VS Code extension to test an API's functionality. You can add test scripts to your collection, folders, and requests. You can also run a collection's or folder's requests in a specific sequence.

### Test your API using scripts

You can add API test scripts to collections, folders, and requests that run after a response is received. Add tests to confirm that your API is working as expected, and help you debug your API if something goes wrong. Postman supports test scripts written in JavaScript.

To add test scripts to collections, folders, and requests, do the following:

1. Select a collection, folder, or request you  const names = [];
  /** @type {Map<string, number>} */
  const sourceIdx = new Map();
  /** @type {Map<string, number>} */
  const nameIdx = new Map();

  /**
   * @param {string | null | undefined} source source identifier
   * @param {string | undefined} content source content (when available)
   * @returns {number} index assigned in the composed map
   */
  const getSourceIdx = (source, content) => {
    const key = source || "";
    let idx = sourceIdx.get(key);
    if (typeof idx === "undefined") {
      idx = sources.length;
      sources.push(key);
      sourcesContent.push(typeof content === "string" ? content : null);
      sourceIdx.set(key, idx);
    } else if (typeof content === "string" && sourcesContent[idx] === null) {
      sourcesContent[idx] = content;
    }
    return idx;
  };

  /**
   * @param {string | null | undefined} value name
   * @returns {number} index assigned in the composed map
   */
  const getNameIdx = value => {
    if (typeof value !== "string") {
      return -1;
    }
    let idx = nameIdx.get(value);
    if (typeof idx === "undefined") {
      idx = names.length;
      names.push(value);
      nameIdx.set(value, idx);
    }
    return idx;
  };
  const decoded = decodedMappings(current);
  const currentSources = current.sources.map(
  /**
   * @param {string | null} source source from current map
   * @returns {string} normalized source string
   */
  source => source || "");
  const currentNames = current.names;

  /** @type {number[][][]} */
  const composed = [];
  for (let line = 0; line < decoded.length; line++) {
    /** @type {number[][]} */
    const newSegments = [];
    for (const rawSeg of decoded[line]) {
      const seg = /** @type {number[]} */rawSeg;

      // Single-element segment is just a generated column with no source info
      if (seg.length < 4) {
        newSegments.push([seg[0]]);
        continue;
      }
      const sourceName = currentSources[seg[1]];
      const origLine = /** @type {number} */seg[2];
      const origCol = /** @type {number} */seg[3];
      const segName = seg.length >= 5 ? currentNames[seg[4]] : (/** @type {string | null} */null);

      // When the segment points back at our intermediate `name`, look up
      // the original position in the previous map and emit a mapping that
      // points all the way back. Otherwise keep the segment as-is.
      if (sourceName === name) {
        const orig = originalPositionFor(previous, {
          line: origLine + 1,
          column: origCol
        });
        if (typeof orig.source !== "string" || orig.line == null || orig.column == null) {
          continue;
        }
        const content = sourceContentFor(previous, orig.source) || undefined;
        const newSrcIdx = getSourceIdx(orig.source, content);
        const finalName = typeof orig.name === "string" && orig.name ? orig.name : segName;
        if (typeof finalName === "string") {
          newSegments.push([seg[0], newSrcIdx, orig.line - 1, orig.column, getNameIdx(finalName)]);
        } else {
          newSegments.push([seg[0], newSrcIdx, orig.line - 1, orig.column]);
        }
      } else {
        const content = sourceContentFor(current, sourceName) || undefined;
        const newSrcIdx = getSourceIdx(sourceName, content);
        if (typeof segName === "string") {
          newSegments.push([seg[0], newSrcIdx, origLine, origCol, getNameIdx(segName)]);
        } else {
          newSegments.push([seg[0], newSrcIdx, origLine, origCol]);
        }
      }
    }
    composed.push(newSegments);
  }
  const result = /** @type {RawSourceMap} */

  /** @type {unknown} */{
    version: 3,
    sources,
    names,
    mappings: encodeMappings(composed)
  };
  if (currentMap.file) {
    result.file = currentMap.file;
  }
  if (sourcesContent.some(value => typeof value === "string")) {
    result.sourcesContent = /** @type {string[]} */
    /** @type {unknown} */sourcesContent;
  }
  return result;
}
/* eslint-enable prefer-destructuring, no-eq-null, eqeqeq */

/**
 * @template T
 * @param {import("./index.js").InternalOptions<T>} options options
 * @returns {Promise<MinimizedResult>} minified result
 */
async function minify(options) {
  const {
    name,
    input,
    inputSourceMap,
    extractComments,
    module,
    ecma
  } = options;
  const {
    implementation,
    options: minimizerOptions
  } = options.minimizer;
  const implementations = Array.isArray(implementation) ? implementation : [implementation];

  /** @type {string | undefined} */
  let lastCode;
  /** @type {RawSourceMap | undefined} */
  let lastMap;
  /** @type {(Error | string)[]} */
  const warnings = [];
  /** @type {(Error | string)[]} */
  const errors = [];
  /** @type {string[]} */
  const extractedComments = [];
  for (let i = 0; i < implementations.length; i++) {
    const currentImplementation = /** @type {import("./index.js").BasicMinimizerImplementation<T> & import("./index.js").MinimizeFunctionHelpers} */
    implementations[i];
    const baseOptions = /** @type {import("./index.js").MinimizerOptions<T> & { module?: boolean, ecma?: number | string }} */

    Array.isArray(minimizerOptions) ? minimizerOptions[i] || {} : minimizerOptions || {};
    const currentInput = typeof lastCode === "string" ? lastCode : input;
    const currentMap = typeof lastCode === "string" ? lastMap : inputSourceMap;

    // Overlay `module` and `ecma` without mutating the caller's options so
    // a single options object can be reused safely across assets.
    const currentOptions = /** @type {import("./index.js").MinimizerOptions<T>} */
    {
      ...baseOptions,
      module: baseOptions.module || module,
      ecma: baseOptions.ecma || ecma
    };
    const result = await currentImplementation({
      [name]: currentInput
    }, currentMap, currentOptions, extractComments);
    if (result.warnings && result.warnings.length > 0) {
      warnings.push(...result.warnings);
    }
    if (result.errors && result.errors.length > 0) {
      errors.push(...result.errors);
    }
    if (result.extractedComments && result.extractedComments.length > 0) {
      extractedComments.push(...result.extractedComments);
    }
    if (typeof result.code === "string") {
      lastCode = result.code;
      // The minimizer's output map is `name → step-output`. Chain it with
      // the previous accumulated map so that across an array of minimizers
      // the final map points back to the original sources.
      lastMap = composeSourceMaps(result.map, currentMap, name);
    }
  }
  return {
    code: lastCode,
    map: lastMap,
    warnings,
    errors,
    extractedComments
  };
}

/**
 * @param {string} options options
 * @returns {Promise<MinimizedResult>} minified result
 */
async function transform(options) {
  // 'use strict' => this === undefined (Clean Scope)
  // Safer for possible security issues, albeit not critical at all here

  const evaluatedOptions =
  /**
   * @template T
   * @type {import("./index.js").InternalOptions<T>}
   */

  // eslint-disable-next-line no-new-func
  new Function("exports", "require", "module", "__filename", "__dirname", `'use strict'\nreturn ${options}`) // eslint-disable-next-line n/exports-style
  (exports, require, module, __filename, __dirname);
  return minify(evaluatedOptions);
}
module.exports = {
  minify,
  transform
};
