// webpack/lib/util/deprecation.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/

"use strict";

const util = require("util");

/** @type {Map<string, () => void>} */
const deprecationCache = new Map();

/**
 * Defines the fake hook marker type used by this module.
 * @typedef {object} FakeHookMarker
 * @property {true} _fakeHook it's a fake hook
 */

/**
 * Defines the shared type used by this module.
 * @template T
 * @typedef {T & FakeHookMarker} FakeHook<T>
 */

/**
 * Creates a deprecation.
 * @param {string} message deprecation message
 * @param {string} code deprecation code
 * @returns {() => void} function to trigger deprecation
 */
const createDeprecation = (message, code) => {
	const cached = deprecationCache.get(message);
	if (cached !== undefined) return cached;
	const fn = util.deprecate(
		() => {},
		message,
		`DEP_WEBPACK_DEPRECATION_${code}`
	);
	deprecationCache.set(message, fn);
	return fn;
};

/** @typedef {"concat" | "entry" | "filter" | "find" | "findIndex" | "includes" | "indexOf" | "join" | "lastIndexOf" | "map" | "reduce" | "reduceRight" | "slice" | "some"} COPY_METHODS_NAMES */

/** @type {COPY_METHODS_NAMES[]} */
const COPY_METHODS = [
	"concat",
	"entry",
	"filter",
	"find",
	"findIndex",
	"includes",
	"indexOf",
	"join",
	"lastIndexOf",
	"map",
	"reduce",
	"reduceRight",
	"slice",
	"some"
];

/** @typedef {"copyWithin" | "entries" | "fill" | "keys" | "pop" | "reverse" | "shift" | "splice" | "sort" | "unshift"} DISABLED_METHODS_NAMES */

/** @type {DISABLED_METHODS_NAMES[]} */
const DISABLED_METHODS = [
	"copyWithin",
	"entries",
	"fill",
	"keys",
	"pop",
	"reverse",
	"shift",
	"splice",
	"sort",
	"unshift"
];

/**
 * Defines the set with deprecated array methods type used by this module.
 * @template T
 * @typedef {Set<T> & { [Symbol.isConcatSpreadable]: boolean } & { push: (...items: T[]) => void, length?: number } & { [P in DISABLED_METHODS_NAMES]: () => void } & { [P in COPY_METHODS_NAMES]: P extends keyof Array<T> ? () => Pick<Array<T>, P> : never }} SetWithDeprecatedArrayMethods
 */

/**
 * Processes the provided set.
 * @template T
 * @param {Set<T>} set new set
 * @param {string} name property name
 * @returns {void}
 */
module.exports.arrayToSetDeprecation = (set, name) => {
	for (const method of COPY_METHODS) {
		if (/** @type {SetWithDeprecatedArrayMethods<T>} */ (set)[method]) continue;
		const d = createDeprecation(
			`${name} was changed from Array to Set (using Array method '${method}' is deprecated)`,
			"ARRAY_TO_SET"
		);
		/** @type {EXPECTED_ANY} */
		(set)[method] =
			// eslint-disable-next-line func-names
			function () {
				d();
				// eslint-disable-next-line unicorn/prefer-spread
				const array = Array.from(this);
				return Array.prototype[
					/** @type {keyof COPY_METHODS} */ (method)
				].apply(
					array,
					// eslint-disable-next-line prefer-rest-params
					arguments
				);
			};
	}
	const dPush = createDeprecation(
		`${name} was changed from Array to Set (using Array method 'push' is deprecated)`,
		"ARRAY_TO_SET_PUSH"
	);
	const dLength = createDeprecation(
		`${name} was changed from Array to Set (using Array property 'length' is deprecated)`,
		"ARRAY_TO_SET_LENGTH"
	);
	const dIndexer = createDeprecation(
		`${name} was changed from Array to Set (indexing Array is deprecated)`,
		"ARRAY_TO_SET_INDEXER"
	);
	/** @type {SetWithDeprecatedArrayMethods<T>} */
	(set).push = function push() {
		dPush();
		// eslint-disable-next-line prefer-rest-params, unicorn/prefer-spread
		for (const item of Array.from(arguments)) {
			this.add(item);
		}
		return this.size;
	};
	for (const method of DISABLED_METHODS) {
		if (/** @type {SetWithDeprecatedArrayMethods<T>} */ (set)[method]) continue;

		/** @type {SetWithDeprecatedArrayMethods<T>} */
		(set)[method] = () => {
			throw new Error(
				`${name} was changed from Array to Set (using Array method '${method}' is not possible)`
			);
		};
	}
	/**
	 * Creates an index getter.
	 * @param {number} index index
	 * @returns {() => T | undefined} value
	 */
	const createIndexGetter = (index) => {
		/**
		 * Returns the value at this location.
		 * @this {Set<T>} a Set
		 * @returns {T | undefined} the value at this location
		 */
		// eslint-disable-next-line func-style
		const fn = function () {
			dIndexer();
			let i = 0;
			for (const item of this) {
				if (i++ === index) return item;
			}
		};
		return fn;
	};
	/**
	 * Define index getter.
	 * @param {number} index index
	 */
	const defineIndexGetter = (index) => {
		Object.defineProperty(set, index, {
			get: createIndexGetter(index),
			set(value) {
				throw new Error(
					`${name} was changed from Array to Set (indexing Array with write is not possible)`
				);
			}
		});
	};
	defineIndexGetter(0);
	let indexerDefined = 1;
	Object.defineProperty(set, "length", {
		get() {
			dLength();
			const length = this.size;
			for (indexerDefined; indexerDefined < length + 1; indexerDefined++) {
				defineIndexGetter(indexerDefined);
			}
			return length;
		},
		set(value) {
			throw new Error(
				`${name} was changed from Array to Set (writing to Array property 'length' is not possible)`
			);
		}
	});
	/** @type {SetWithDeprecatedArrayMethods<T>} */
	(set)[Symbol.isConcatSpreadable] = true;
};

/**
 * Returns } SetDeprecatedArray.
 * @template T
 * @param {string} name name
 * @returns {{ new <T = EXPECTED_ANY>(values?: ReadonlyArray<T> | null): SetDeprecatedArray<T> }} SetDeprecatedArray
 */
module.exports.createArrayToSetDeprecationSet = (name) => {
	let initialized = false;

	/**
	 * Represents SetDeprecatedArray.
	 * @template T
	 */
	class SetDeprecatedArray extends Set {
		/**
		 * Creates an instance of SetDeprecatedArray.
		 * @param {ReadonlyArray<T> | null=} items items
		 */
		constructor(items) {
			super(items);
			if (!initialized) {
				initialized = true;
				module.exports.arrayToSetDeprecation(
					/** @type {SetWithDeprecatedArrayMethods<T>} */
					(SetDeprecatedArray.prototype),
					name
				);
			}
		}
	}
	return SetDeprecatedArray;
};

/**
 * Returns fake hook which redirects.
 * @template {object} T
 * @param {T} fakeHook fake hook implementation
 * @param {string=} message deprecation message (not deprecated when unset)
 * @param {string=} code deprecation code (not deprecated when unset)
 * @returns {FakeHook<T>} fake hook which redirects
 */
module.exports.createFakeHook = (fakeHook, message, code) => {
	if (message && code) {
		fakeHook = deprecateAllProperties(fakeHook, message, code);
	}
	return Object.freeze(
		Object.assign(fakeHook, { _fakeHook: /** @type {true} */ (true) })
	);
};

/**
 * Deprecate all properties.
 * @template T
 * @param {T} obj object
 * @param {string} message deprecation message
 * @param {string} code deprecation code
 * @returns {T} object with property access deprecated
 */
const deprecateAllProperties = (obj, message, code) => {
	const newObj = {};
	const descriptors = Object.getOwnPropertyDescriptors(obj);
	for (const name of Object.keys(descriptors)) {
		const descriptor = descriptors[name];
		if (typeof descriptor.value === "function") {
			Object.defineProperty(newObj, name, {
				...descriptor,
				value: util.deprecate(descriptor.value, message, code)
			});
		} else if (descriptor.get || descriptor.set) {
			Object.defineProperty(newObj, name, {
				...descriptor,
				get: descriptor.get && util.deprecate(descriptor.get, message, code),
				set: descriptor.set && util.deprecate(descriptor.set, message, code)
			});
		} else {
			let value = descriptor.value;
			Object.defineProperty(newObj, name, {
				configurable: descriptor.configurable,
				enumerable: descriptor.enumerable,
				get: util.deprecate(() => value, message, code),
				set: descriptor.writable
					? util.deprecate(
							/**
							 * Handles the callback logic for this hook.
							 * @template T
							 * @param {T} v value
							 * @retur)   s h o u l d   a l s o   p r o v i d e   t h e   p r i m a r y   s t a t u s   o f   " 
 
                     " t h e   e l e m e n t . "   ) ,   
 
               V a l u e M a p   {   " 0 " ,   " 1 " ,   " 2 " ,   " 3 " ,   " 4 " ,   " 5 " ,   " 6 " ,   " 7 " ,   " 8 " ,   " 9 " ,   
 
                     " 1 0 " ,   " 1 1 " ,   " 1 2 " ,   " 1 3 " ,   " 1 4 " ,   " 1 5 " ,   " 1 6 " ,   " 1 7 " ,   " 1 8 " ,   
 
                     " . . " ,   " 0 x 8 0 0 0 . . "   } ,   
 
               V a l u e s   {   " U n k n o w n " ,   " O t h e r " ,   " O K " ,   " D e g r a d e d " ,   " S t r e s s e d " ,   
 
                     " P r e d i c t i v e   F a i l u r e " ,   " E r r o r " ,   " N o n - R e c o v e r a b l e   E r r o r " ,   
 
                     " S t a r t i n g " ,   " S t o p p i n g " ,   " S t o p p e d " ,   " I n   S e r v i c e " ,   
 
                     " N o   C o n t a c t " ,   " L o s t   C o m m u n i c a t i o n " ,   " A b o r t e d " ,   " D o r m a n t " ,   
 
                     " S u p p o r t i n g   E n t i t y   i n   E r r o r " ,   " C o m p l e t e d " ,   " P o w e r   M o d e " ,   
 
                     " D M T F   R e s e r v e d " ,   " V e n d o r   R e s e r v e d "   } ,   
 
               A r r a y T y p e   (   " I n d e x e d "   ) ,   
 
               M o d e l C o r r e s p o n d e n c e   {   
 
                     " C I M _ M a n a g e d S y s t e m E l e m e n t . S t a t u s D e s c r i p t i o n s "   } ] 
 
       u i n t 1 6   O p e r a t i o n a l S t a t u s [ ] ; 
 
 
 
             [ D e s c r i p t i o n   (   
 
                     " S t r i n g s   d e s c r i b i n g   t h e   v a r i o u s   O p e r a t i o n a l S t a t u s   a r r a y   " 
 
                     " v a l u e s .   F o r   e x a m p l e ,   i f   \ " S t o p p i n g \ "   i s   t h e   v a l u e   " 
 
                     " a s s i g n e d   t o   O p e r a t i o n a l S t a t u s ,   t h e n   t h i s   p r o p e r t y   m a y   " 
 
                     " c o n t a i n   a n   e x p l a n a t i o n   a s   t o   w h y   a n   o b j e c t   i s   b e i n g   " 
 
                     " s t o p p e d .   N o t e   t h a t   e n t r i e s   i n   t h i s   a r r a y   a r e   c o r r e l a t e d   " 
 
                     " w i t 
